src/consensus/read_validator.rs

```diff
@@ -96,7 +98,41 @@ impl ReadValidator {
         }
     }
 
+    fn verify_signatures(value: &proto::DecidedValue) -> bool {
+        let certificate = match value.value.as_ref().unwrap() {
+            proto::decided_value::Value::Shard(shard_chunk) => shard_chunk
+                .commits
+                .as_ref()
+                .unwrap()
+                .to_commit_certificate(),
+
+            proto::decided_value::Value::Block(block) => {
+                block.commits.as_ref().unwrap().to_commit_certificate()
+            }
+        };
+
+        for signature in certificate.aggregated_signature.signatures {
+            let vote = Vote::new_precommit(
+                certificate.height,
+                certificate.round,
+                NilOrVal::Val(certificate.value_id.clone()),
+                signature.address.clone(),
+            );
+
+            let public_key = PublicKey::try_from_bytes(&signature.address.0).unwrap();
+            if !public_key.verify(&vote.to_sign_bytes(), &signature.signature.0) {
```

Comment:
We need to add one more check. This should use your validator set config from your other PR, and ensure that the public keys are correct for the height being verified.

Reply:
Done!